Made some slight improvements to the AssetBrowser

penev92 opened the pull request:
This PR adds a few simple improvements to the AssetBrowser. Commit messages should be pretty explanatory. I strongly advise reviewing commit by commit.

P.S.:
Fixes #19187.

penev92 committed:
Paused shellmap music when opening videos in the AssetBrowser

penev92 committed:
Added a scale slider for models in the AssetBrowser

penev92 committed:
Added a scale slider for sprites in the AssetBrowser

pchote commented on OpenRA.Mods.Common/AssetBrowser.cs:
```diff
@@ -13,7 +13,8 @@ namespace OpenRA
 	{
-		[FieldLoader.Require]
-		public readonly string[] SupportedExtensions = new string[0];
+		public readonly string[] SpriteExtensions = new string[0];
```
`= { };`?

penev92 replied:
Is there a meaningful difference?

pchote commented on OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs:
```diff
@@ -402,18 +402,26 @@ bool LoadAsset(IReadOnlyPackage package, string filename)
 					if (video != null)
 					{
+						// Pause music so it doesn't interfere with current asset.
+						Game.Sound.PauseMusic();
```
Can we please instead implement this the same way `MissionBrowserLogic` does it?

pchote commented on OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs:
```diff
@@ -267,6 +270,17 @@ public AssetBrowserLogic(Widget widget, Action onExit, ModData modData, World wo
+			var modelScaleSlider = panel.GetOrNull<SliderWidget>("MODEL_SCALE_SLIDER");
+			if (modelScaleSlider != null)
+			{
+				modelScaleSlider.OnChange += x =>
```
`modelScaleSlider.OnChange += x => modelScale = x;`?

pchote commented on OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs:
```diff
@@ -230,6 +233,19 @@ public AssetBrowserLogic(Widget widget, Action onExit, ModData modData, World wo
+			var spriteScaleSlider = panel.GetOrNull<SliderWidget>("SPRITE_SCALE_SLIDER");
+			if (spriteScaleSlider != null)
+			{
+				spriteScaleSlider.OnChange += x =>
```
`spriteScaleSlider.OnChange += x => spriteScale = x;`?

pchote commented on OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs:
```diff
@@ -390,6 +394,23 @@ void AddAsset(ScrollPanelWidget list, string filepath, IReadOnlyPackage package,
+				currentFrame = 0;
+			}
+
+			if (currentVoxel != null)
```
There is no harm in setting these to null if they are already null, so how about just
```csharp
if (currentSound != null)
	Game.Sound.StopSound(currentSound);

currentSprites = null;
currentFrame = 0;
currentVoxel = null;
currentSound = null;
```
?

pchote commented on OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs:
```diff
@@ -442,9 +463,19 @@ bool LoadAsset(IReadOnlyPackage package, string filename)
+					// Pause music so it doesn't interfere with current asset.
+					Game.Sound.PauseMusic();
+
+					using (var soundStream = Game.ModData.DefaultFileSystem.Open(filename))
```
This needs to be `prefix + filename`, otherwise the voices in TS's speech-gdi end up playing the nod versions.

pchote reviewed (commented):
Thanks for splitting this out into individual commits, it indeed made the review much easier.

I can also reliably trigger a crash by selecting an audio file while a sprite animation is playing.

penev92 commented:
Updated as requested.

penev92 commented:
Question for the audience: why do we even have `AssetBrowser.Supported*Extensions` instead of using what the mod already defines with `SoundFormats`, `SpriteFormats` and `VideoFormats`?  🤔

reaperrr commented:
Clicking on a sound file in the asset browser just directly plays it, could we get Play/Stop buttons instead, like for VQAs?
Playback and extensions as such work fine, though.

> Question for the audience: why do we even have AssetBrowser.Supported*Extensions instead of using what the mod already defines with SoundFormats, SpriteFormats and VideoFormats?

Only a guess, but since multiple extensions can be used for the same format (.tem and .sno for shps etc.), it's probably more efficient to pre-filter by extensions, rather than checking every files' header or something.
On that note, RA and TD asset browsers are missing the .V00-V03 extensions for the voice report variations, as well as the .VAR extension for the alternate music and sound files in original TD mix files.

pchote commented:
> Clicking on a sound file in the asset browser just directly plays it, could we get Play/Stop buttons instead, like for VQAs?

The bigger problem (with the same fix) is that the sounds keeps on restarting when you are typing into the search filter.

penev92 committed:
Fixed AudFormat and WavFormat implementations of ISoundFormat.LengthInSeconds

- Fixed a rounding issue in `WavReader.WaveLength()`.
- Fixed `AudReader.SoundLength()` not resetting the stream position.
- Fixed crashes caused by disposed streams because `LengthInSeconds` would try and calculate the length on the fly. It is now precalculated and cached (making it consistent across all 5 current `ISoundFormat` implementations).
- Fixed a crash in `AudReader.LoadSound()`'s `out Func<Stream> result` because that func would try and access the disposed stream's `Length` property. That works for `SegmentStream`, but not for `FileStream`.
- Fixed frameCount/soundLength label positioning in the AssetBrowser window to avoid text clipping .

penev92 commented:
> Only a guess, but since multiple extensions can be used for the same format (.tem and .sno for shps etc.), it's probably more efficient to pre-filter by extensions, rather than checking every files' header or something.
On that note, RA and TD asset browsers are missing the .V00-V03 extensions for the voice report variations, as well as the .VAR extension for the alternate music and sound files in original TD mix files.

My proposal would be to go from the current (at least in this PR) setup:
```yaml
SoundFormats: Aud, Wav

SpriteFormats: ShpTS, TmpTS, ShpTD

VideoFormats: Vqa

ModelSequenceFormat: VoxelModelSequence

AssetBrowser:
	SpriteExtensions: .shp, .tem, .sno
	ModelExtensions: .vxl
	AudioExtensions: .aud, .wav
	VideoExtensions: .vqa
```

to the `AssetBrowser` node being obsolete by giving each `*Loader` class a list of supported extensions (hardcoded, possibly, but could be made extensible?) and the asset browser querying `SoundFormats`/`SpriteFormats`/etc. directly.
That then got me thinking we can do one better and condense it all into 
```yaml
SupportedFileFormats:
	Sprites: ShpTS, TmpTS, ShpTD
	Models: Vxl # or `Voxel` or `VoxelModelSequence`
	Audio: Aud, Wav
	Video: Vqa
```
to be used by both engine and asset browser alike while maintaining the same functionality that we have with this PR. (ofc the proposal would be left for a separate, possibly follow-up PR).

penev92 commented:
Disabled auto play of audio files and enabled the Play and Stop buttons.
Doing that exposed issues in both `AudFormat` and `WavFormat`'s `ISoundFormat.LengthInSeconds` implementations (via `AudReader.SoundLength()` and `WavReader.WaveLength()`), so I fixed those as well.
During all this I also noticed an incorrectly installed D2k song and fixed the installers, although that commit (the last one) might want to go to its own PR.

pchote commented:
>  During all this I also noticed an incorrectly installed D2k song and fixed the installers, although that commit (the last one) might want to go to its own PR.

Own PR that we can pick to prep, IMO.

penev92 commented:
Split out the D2k installers fix into #19176 and removed that commit from this branch.

penev92 commented:
> I can also reliably trigger a crash by selecting an audio file while a sprite animation is playing.

Looks like that is also an issue on `bleed` when you select a voxel while a sprite animation is playing - #19187 😉  That should already be fixed in this PR.

teinarss commented on OpenRA.Mods.Cnc/FileFormats/AudReader.cs:
```diff
@@ -34,6 +34,8 @@ public static class AudReader
 	{
 		public static float SoundLength(Stream s)
 		{
+			var position = s.Position;
```
originalPosition?

pchote commented:
Sounds aren't stopped when you leave the close the asset browser, so if you start playing a song and then cancel you end up with two songs playing at once!

IceReaper commented:
My suggestions:
- Add scrolling to the viewport, when viewing assets larger than the preview space (OpenKrush maps are huge images and way bigger for example)
- When multiple assets have the same filename, selecting one will always display the first one. Entries should point explicit to its package (Several old games have no filenames, so "0.sprite" might be present multiple times)
- Can we add palette support? Basicaly allow the asset browser to load up palettes from assets, and make them selectable without the need to write rules for those. (When implementing a new RE'd game, you might want to avoid writing a huge rules block. Being able to test found palettes might be a nice addition.
- Can we re-layout the asset browser so its inner bounds are tied to its outer bounds? In other words: If we make the outer window 100% width/height, every should properly stretch to the size, instead of sitting in the top left. (did this for OpenKrush, to display larger assets. Palette and close button fit good, everything else is tied to top left.
- Can we add a tooltip to the asset list showing its package? (When having multiple same-named files, you want to see which one it is, and who contains it, avoids the need to manualy cycle through all the packages)

penev92 committed:
Split AssetBrowser supported formats

Splitting them from one array into separate allows us to then reliably pick how each asset should be presented. Also lets us unhardcode some checks like "if file is .vxl ... else is sprite".

penev92 committed:
Added manual playing and stopping of audio in the AssetBrowserLogic

Also added visualisation of audio file length and progress and improved asset cleanup.

penev92 committed:
Added containing package name to asset tooltip

penev92 commented:
Rebased. Couldn't spot any issues with audio. Added one more commit to add package name to the asset tooltip as @IceReaper requested.
The other requests by @IceReaper are out of scope of what should have been a pretty simple PR. (but the "multiple assets with the same name" issue comes from the fact that for sprites and models we don't load the files, but rather query `SpriteCache`/`ModelCache` by asset name).

Smittytron commented:
I didn't notice any issues with audio; was testing with music playing. The audio side of this is a great help to my scripting work when I'm trying to find the right sound!

The icon .shp's don't have a slider, but are scaled accordingly when you use the slider on something else and then go back. I don't know if you want to lock them at an appropriate size or apply the scaler to them.

TS defender at full scale is quite funny. All the other TS browser issues I saw are present in bleed.

Mailaender commented on OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs:
```diff
@@ -239,7 +274,16 @@ public AssetBrowserLogic(Widget widget, Action onExit, ModData modData, WorldRen
+				spriteScaleSlider.OnChange += x => spriteScale = x;
+				spriteScaleSlider.GetValue = () => spriteScale;
+				spriteScaleSlider.IsVisible = () => !isVideoLoaded && currentSoundFormat == null;
+				panel.GetOrNull<LabelWidget>("SPRITE_SCALE").IsVisible = () => !isVideoLoaded && currentSoundFormat == null;
```
Icons don't get a scale slider:

![Screenshot_20210917_203506](https://user-images.githubusercontent.com/756669/133837634-c2b716c8-e1c8-47cf-bdf2-9c6c893655e1.png)

This might already be fixed by https://github.com/OpenRA/OpenRA/pull/19498.

Mailaender commented on OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs:
```diff
@@ -397,41 +452,75 @@ bool LoadAsset(IReadOnlyPackage package, string filename)
+					var soundStream = Game.ModData.DefaultFileSystem.Open(prefix + filename);
+
+					foreach (var modDataSoundLoader in Game.ModData.SoundLoaders)
+						if (modDataSoundLoader.TryParseSound(soundStream, out currentSoundFormat))
```
Skipping the braces here is a bit too much imho.

abcdefg30 replied:
We require braces for the outer `for` if the inner block uses braces.

Mailaender reviewed (dismissed):
This is not a small, but a huge improvement. Especially the upscaling is so important when inspecting pixelart. Thanks!

Mailaender commented on OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs:
```diff
@@ -397,41 +452,75 @@ bool LoadAsset(IReadOnlyPackage package, string filename)
+					}
+				}
+				else
+				{
```
You can skip brackets here.

penev92 commented:
Fixed the braces.
Icons don't get a scale slider because that's inside the container holding the other sprite controls and that is designed to hide when the current sprite only has one frame. There is no trivial way around that so I'd rather skip fixing this at this time.

abcdefg30 commented on OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs:
```diff
@@ -397,42 +452,76 @@ bool LoadAsset(IReadOnlyPackage package, string filename)
 
-					if (frameSlider != null)
+					// For whatever reason disposing of this stream immediately still lets sound play in all mods except D2k.
+					var soundStream = Game.ModData.DefaultFileSystem.Open(prefix + filename);
```
Does this mean we don't dispose those streams because of D2k?

penev92 replied:
Apparently I hadn't done enough digging on this. It is because D2k uses `FileStream` for its AUD files (the music is unpacked), while everything else (music in other mods + other audio) is packed in container files and uses `SegmentStream`. There seems to be a difference in how the two stream implementations are disposed. `SegmentStream` keeps its `Length` even after being disposed, but `FileStream` does not. That leads to a crash in `AudReader.LoadSound()`'s `out Func<Stream> result` when we try to use it later. The fix is actually trivial and contained to `AudReader`.
(I'm not 100% this is what I meant with that comment, but adding a `using` to the stream made D2k music crash, so I traced it to 👆. Everything should work now even with disposing the stream. )

Fixed and removed the comment.

abcdefg30 commented on OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs:
```diff
@@ -540,5 +629,47 @@ string GetSourceDisplayName(IReadOnlyPackage source)
+			if (currentSound != null)
+				Game.Sound.StopSound(currentSound);
+
+			if (currentAudioStream != null)
+			{
+				currentAudioStream.Dispose();
+				currentAudioStream = null;
+			}
```
You should be able to simplify this to
```suggestion
			currentAudioStream?.Dispose();
			currentAudioStream = null;
```

abcdefg30 commented on OpenRA.Mods.Common/Widgets/SpriteWidget.cs:
```diff
@@ -61,7 +63,7 @@ public override void Draw()
 
 			if (sprite != cachedSprite)
```
This should probably get a `|| scale != cachedScale` and be combined with the if below to reduce duplication.

abcdefg30 commented on OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs:
```diff
@@ -397,42 +452,76 @@ bool LoadAsset(IReadOnlyPackage package, string filename)
+					currentVoxel = null;
 
-				if (Path.GetExtension(filename.ToLowerInvariant()) == ".vxl")
+					// Just in case we're switching away from a type of asset that forced the music to pause.
+					UnMuteSounds();
```
Wouldn't it make more sense to do this when "unloading" an asset?

abcdefg30 replied:
Or rather, why don't we do this at the top while cleaning the old asset? I assume that is because the cases below mute right again, but those calls are there anyway. I'd prefer unmuting above for visibility/generality. What do you think?

penev92 replied:
Moved directly to `ClearLoadedAssets()`.

penev92 commented:
Fixed the raised style issues.
Investigating the audio stream disposal revealed a few other possible things to change to fix bugs or just add general polish.
Changes were mostly in commits `Fixed AudFormat and WavFormat implementations of ISoundFormat.LengthInSeconds` and `Added manual playing and stopping of audio in the AssetBrowserLogic`.

penev92 commented:
I suppose this may get asked, so I'd rather have it sooner rather than later: is @Mailaender's +1 still valid, because there were a few more fixes since then?

RoosterDragon commented on OpenRA.Mods.Common/FileFormats/WavReader.cs:
```diff
@@ -123,7 +123,7 @@ public static float WaveLength(Stream s)
 			var length = s.Length * 8;
 
-			return length / (channels * sampleRate * bitsPerSample);
+			return (float)length / (channels * sampleRate * bitsPerSample);
```
You're resetting the stream position for `AudReader.SoundLength` but not here in `WavReader.WaveLength` - an error?

You could consider removing the separate `*Length` methods and return another out parameter from the `LoadSound` methods instead - this would remove faff in co-ordinating the stream position between methods.

RoosterDragon commented on OpenRA.Mods.Common/AssetBrowser.cs:
```diff
@@ -13,7 +13,9 @@ namespace OpenRA
+		public readonly string[] SpriteExtensions = { };
+		public readonly string[] ModelExtensions = { };
+		public readonly string[] AudioExtensions = { };
+		public readonly string[] VideoExtensions = { };
```
Consider `Array.Empty<string>()`.

RoosterDragon commented on OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs:
```diff
@@ -45,11 +49,16 @@ public class AssetBrowserLogic : ChromeLogic
 		IModel currentVoxel;
+		ISound currentSound;
+		ISoundFormat currentSoundFormat;
+		Stream currentAudioStream;
```
`currentAudioStream` is unused

penev92 replied:
Brought it back in one of the new commits.

RoosterDragon commented on OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs:
```diff
@@ -397,42 +456,70 @@ bool LoadAsset(IReadOnlyPackage package, string filename)
-						frameSlider.Ticks = currentSprites.Length;
+						foreach (var modDataSoundLoader in Game.ModData.SoundLoaders)
+						{
+							if (modDataSoundLoader.TryParseSound(soundStream, out currentSoundFormat))
```
It's somewhat cheeky to rely on `currentSoundFormat` having working methods after being disposed here, however I'm willing to live with it.

penev92 replied:
I put a bunch of effort into making the cheekiness work for AUD and WAV, but apparently it still breaks on OGG and MP3 (no idea about VOC atm), so will need to find another way I guess.

penev92 replied:
Ultimately brought back `currentAudioStream` as to not have to dispose the stream prematurely. Hopefully nothing bad comes out of that.

RoosterDragon reviewed (changes requested):
The scale slider can push some items outside the intended area - would it be possible to clip them instead for a bit of polish?
![AssetScale](https://user-images.githubusercontent.com/3399086/141675015-44656cb3-5487-4d37-ad5d-4266f34cf207.gif)

penev92 committed:
Reworked ISoundFormat.LengthInSeconds implementations

penev92 committed:
Fixed background music playing momentarily in AssetBrowser when switching to a video asset

penev92 committed:
Fixed AssetBrowserLogic disposing audio streams prematurely

penev92 committed:
Fixed Mp3Loader and OggLoader not resetting stream position

penev92 committed:
Added audio playback to the AssetBrowser

penev92 commented:
Rebased.
That means you can now test from the comfort of your .NET 6 environments 😎 
Also ran into trouble with the changes from #19681 so will need help with the TD asset browser. @dragunoff 
Fixed a bunch more things, including a couple of issues in OggLoader and Mp3Loader. The notable new changes are in new commits, while several not mention-worthy changes have been fixed in the original commits.

Sadly this has turned from "simple AssetBrowser improvements" to mostly an audio parser refactor, but I hope we're finally good to go.

RoosterDragon commented on OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs:
```diff
@@ -397,42 +456,68 @@ bool LoadAsset(IReadOnlyPackage package, string filename)
+					MuteSounds();
 
-					if (frameSlider != null)
+					currentAudioStream = Game.ModData.DefaultFileSystem.Open(prefix + filename);
```
Presumably we need to dispose any existing instance assigned to `currentAudioStream ` before reassigning?

penev92 replied:
That is already handled by `LoadAsset` (the current method) calling `ClearLoadedAssets` as the first thing it does. (Please correct me if I'm wrong ... after 1625372516 rebases and fixes there's a non-zero possibility I may have messed something up again)

penev92 committed:
Updated scale sliders to accout for the new AssetBrowser layout

penev92 committed:
Added hiding of the palette picker in the AssetBrowser

penev92 commented:
Updated the scale sliders position. Also added hiding of the palette controls.

Mailaender commented:
Thanks. Especially, the mixing of different sound formats without a crash is a huge benefit to modders.

Mailaender commented:
[ModSDK Changes](https://github.com/OpenRA/OpenRAModSDK/wiki/Update-notes:-Release-20210321-to-next-release/_compare/bc363c71915a9ccbd0a437698163b6b56ce5996f...62efe6ade0241dc12786406f8fd8f47d63148caf)
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/21a1d7368f0f52c2d3f0f6de8360e989341bc35a...fea425e2993120d25ded7ac36444962e5af59a2f)

Edit by @penev92:
[More Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/bb30b847fc871b5e09332de703f21e96f863ed11...9049d0937d177dd9c54566cb3aef5a553062e2da)
